Fix left division for Hermitian sparse array and transpose/adjoint rhs

goggle opened the pull request:
This partly addresses JuliaLang/LinearAlgebra.jl#538.
The other mentioned issues in this thread are still unsolved, so JuliaLang/LinearAlgebra.jl#538 should either remain open or the other mentioned issues should get their own tickets, even if this gets merged.

goggle committed:
Fix left division for Hermitian sparse array and tranpose/adjoint rhs

dkarrasch reviewed (changes requested):
This is a good start, but the tests are probably tricky and require some more work, to really test the added methods.

goggle committed:
Use `eachindex` to iterate through matrix

goggle committed:
Remove space

goggle committed:
Improve tests

Make sure we test the newly added methods, even if the algorithm for
`\` in `linalg.jl` changes in the future.

goggle commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -804,7 +804,15 @@ function Dense{T}(A::StridedVecOrMat) where T<:VTypes
+    d = allocate_dense(size(A, 1), size(A, 2), size(A, 1), T)
+    s = unsafe_load(pointer(d))
+    for i in 1:length(A)
+        unsafe_store!(s.x, A[i], i)
```
I have modified this code, so that it uses `eachindex` now. As far as I understand this, it should not depend on being one-indexed anymore.

dkarrasch replied:
I was actually thinking that it does, via `unsafe_store!`, and that the same applies to the method above that you haven’t touched. In that sense, I thought your approach was „more correct“ than what we have above. But I’m not familiar with `unsafe_store!` and `unsafe_load`.

goggle replied:
In the initial version
```julia
for i in 1:length(A)
    unsafe_store!(s.x, A[i], i)
end
```
I have assumed that the array `A` is 1-indexed, because I have accessed `A` (inside the loop) at `1`, `2`, ..., `length(A)`.

In the current version
```julia
for (i, c) in enumerate(eachindex(A))
    unsafe_store!(s.x, A[c], i)
end
```
I access `A` at indices given by the `eachindex` function, so they should be valid, even if `A` is not 1-indexed.

dkarrasch replied:
I understand all that. But assume that the first index is `-100`. Then I’m not sure that `unsafe_store!` will work correctly, at least judging by its docstring.

goggle replied:
It seems to work fine for such cases:
```julia
julia> using OffsetArrays

julia> import SuiteSparse.CHOLMOD: allocate_dense

julia> A = OffsetArray{Float64}(undef, -100:-95, -100:-95)
       for (i,c) in enumerate(eachindex(A))
            A[c] = i
       end

julia> A
6×6 OffsetArray(::Array{Float64,2}, -100:-95, -100:-95) with eltype Float64 with indices -100:-95×-100:-95:
 1.0   7.0  13.0  19.0  25.0  31.0
 2.0   8.0  14.0  20.0  26.0  32.0
 3.0   9.0  15.0  21.0  27.0  33.0
 4.0  10.0  16.0  22.0  28.0  34.0
 5.0  11.0  17.0  23.0  29.0  35.0
 6.0  12.0  18.0  24.0  30.0  36.0

julia> B = A'
6×6 LinearAlgebra.Adjoint{Float64,OffsetArray{Float64,2,Array{Float64,2}}} with indices -100:-95×-100:-95:
  1.0   2.0   3.0   4.0   5.0   6.0
  7.0   8.0   9.0  10.0  11.0  12.0
 13.0  14.0  15.0  16.0  17.0  18.0
 19.0  20.0  21.0  22.0  23.0  24.0
 25.0  26.0  27.0  28.0  29.0  30.0
 31.0  32.0  33.0  34.0  35.0  36.0

julia> d = allocate_dense(6, 6, 6, Float64)
       GC.@preserve d begin
           s = unsafe_load(pointer(d))
           for (i,c) in enumerate(eachindex(B))
               unsafe_store!(s.x, B[c], i)
           end
       end

julia> d
6×6 SuiteSparse.CHOLMOD.Dense{Float64}:
  1.0   2.0   3.0   4.0   5.0   6.0
  7.0   8.0   9.0  10.0  11.0  12.0
 13.0  14.0  15.0  16.0  17.0  18.0
 19.0  20.0  21.0  22.0  23.0  24.0
 25.0  26.0  27.0  28.0  29.0  30.0
 31.0  32.0  33.0  34.0  35.0  36.0
```
@dkarrasch Do you think further changes are needed here?

goggle commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -1716,7 +1737,7 @@ function \(A::RealHermSymComplexHermF64SSL, B::StridedVecOrMat)
     end
 end
-function \(adjA::Adjoint{<:Any,<:RealHermSymComplexHermF64SSL}, B::StridedVecOrMat)
+function \(adjA::Adjoint{<:Any, <:RealHermSymComplexHermF64SSL}, B::StridedVecOrMatInclAdjAndTrans)
```
I have reverted this space; didn't intended to modify it in the first place :smile: .

goggle commented on stdlib/SuiteSparse/test/cholmod.jl:
```diff
@@ -855,3 +855,18 @@ end
 end
+
+@testset "Issue #27860" begin
+    A = sparse(Float64[2.0 0.1; 0.1 2.0])
```
I have modified the tests according to your sketch above. But I didn't understand, what you wanted to show with `m`, so I left it out for now. In your example, `m` is `1 x 1` but the RHS is `2 x 1` or `2 x 2`, so the dimensions do not match...

dkarrasch replied:
The rhs’s for `m` (or `a`) are transposes of vectors, so they are row vectors (1x2). I don’t see any other way of testing the `Adjoint{<:Any,<:Vector}` methods, of which I’m not so sure if we really need them anyway.

goggle commented:
Thanks for the review @dkarrasch. I have made some changes, feel free to have a look at it again.

dkarrasch commented on stdlib/SuiteSparse/test/cholmod.jl:
```diff
@@ -855,3 +855,13 @@ end
+    for typeA in (Float64, ComplexF64), typeB in (Float64, ComplexF64), transform in (adjoint, transpose)
+        A = sparse(typeA[2.0 0.1; 0.1 2.0])
+        B = randn(typeB, 2, 2)
+        b = randn(typeB, 1, 2)
```
This is a `Matrix`, and `transform(b)` is an `Adjoint/Transpose{<:Any,<:AbstractMatrix}`, not `Adjoint/Transpose{<:Any,<:AbstractVector}`.

dkarrasch replied:
With the `Vector` methods deleted, I’m no longer sure the `b` tests add anything. At least, we don’t seem to test previously untested methods.

yuyichao commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -804,7 +804,15 @@ function Dense{T}(A::StridedVecOrMat) where T<:VTypes
-function Dense(A::StridedVecOrMat)
+function Dense{T}(A::Union{Adjoint{<:Any, <:StridedVecOrMat}, Transpose{<:Any, <:StridedVecOrMat}}) where T<:VTypes
+    d = allocate_dense(size(A, 1), size(A, 2), size(A, 1), T)
+    s = unsafe_load(pointer(d))
```
`GC.@preserve d` until the function returns.

goggle committed:
Remove adoint/tranpose vector methods

goggle committed:
Add GC.@preserve d

yuyichao commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -806,7 +806,7 @@ function Dense{T}(A::StridedVecOrMat) where T<:VTypes
 function Dense{T}(A::Union{Adjoint{<:Any, <:StridedVecOrMat}, Transpose{<:Any, <:StridedVecOrMat}}) where T<:VTypes
     d = allocate_dense(size(A, 1), size(A, 2), size(A, 1), T)
-    s = unsafe_load(pointer(d))
+    GC.@preserve d s = unsafe_load(pointer(d))
```
You need to preserve d "until the function returns." since you are still using the pointer associated with `d`.
The function above also seems to be missing it.

goggle committed:
Use `GC.@preserve` until function returns

dkarrasch commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -798,13 +798,25 @@ get_perm(FC::FactorComponent) = get_perm(Factor(FC))
+    GC.@preserve d begin
+        s = unsafe_load(pointer(d))
+        for (i, c) in enumerate(eachindex(A))
+            unsafe_store!(s.x, A[c], i)
```
Ok, I misread that earlier. It works, because the `i`s run from 1 to `length(A)`, and not through the indices of `A` (like `c`). I wonder if we should take the same approach in other method above, I mean with `enumerate(eachindex(A))` etc.

goggle committed:
Remove unneeded test cases

goggle committed:
Adopt `enumerate` scheme to already existing `Dense` constructor

dkarrasch commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -798,13 +798,25 @@ get_perm(FC::FactorComponent) = get_perm(Factor(FC))
+        end
+    end
+    d
+end
```
It seems to me that these two methods are exactly the same. So you could introduce
```julia
const StridedVecOrMatInclAdjAndTrans = Union{StridedVecOrMat, Adjoint{<:Any, <:StridedVecOrMat}, Transpose{<:Any, <:StridedVecOrMat}}
```
around here, and have the method as
```julia
function Dense{T}(A::StridedVecOrMatInclAdjAndTrans) where T<:VTypes
```

dkarrasch reviewed (changes requested):
Sorry for being nitpicking, but I just realized one more thing. Everything else looks great.

goggle committed:
Combine `Dense{T}` methods into one

goggle commented:
@dkarrasch No need to apologize, I appreciate to have a careful review! Maintaining good code quality is important IMO.

dkarrasch commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -796,15 +796,18 @@ get_perm(FC::FactorComponent) = get_perm(Factor(FC))
-        unsafe_store!(s.x, A[i], i)
+const StridedVecOrMatInclAdjAndTrans = Union{StridedVecOrMat, Adjoint{<:Any, <:StridedVecOrMat}, Transpose{<:Any, <:StridedVecOrMat}}
+function Dense{T}(A::StridedVecOrMatInclAdjAndTrans) where T<:VTypes
+    d = allocate_dense(size(A, 1), size(A, 2), size(A, 1), T)
```
This is outside of my area of expertise, but I would rather not change the `stride(A, 2)` we had before to `size(A, 1)`. I guess it was written for a reason.

goggle replied:
Yeah, that was that small difference between the two `Dense{T}` methods, that still existed. I cannot use `stride(A, 2)` here, because `stride` is not defined for `Adjoint` or `Transpose`... So I see 2 possibilities:

1. Revert the last change and use two `Dense{T}` methods again.
2. Keep it as it is.

What should I do?

dkarrasch replied:
I thought I tested it because I wasn't sure myself, but I must have messed it up. I now see that there is no `stride` function for `Adjoint`s and `Transpose`s. Sorry for that. Then I'd suggest to have two methods again, just to make sure we don't change behavior of existing methods unnecessarily (beyond the `GC.@preserve` stuff).

dkarrasch commented on stdlib/SuiteSparse/src/cholmod.jl:
```diff
@@ -1703,7 +1712,8 @@ const RealHermSymComplexHermF64SSL = Union{
     Hermitian{Float64,SparseMatrixCSC{Float64,SuiteSparse_long}},
     Hermitian{Complex{Float64},SparseMatrixCSC{Complex{Float64},SuiteSparse_long}}}
-function \(A::RealHermSymComplexHermF64SSL, B::StridedVecOrMat)
+const StridedVecOrMatInclAdjAndTrans = Union{StridedVecOrMat, Adjoint{<:Any, <:StridedVecOrMat}, Transpose{<:Any, <:StridedVecOrMat}}
```
With the above definition, this one is obsolete now, isn't it?

goggle committed:
Remove duplicate definition

goggle committed:
Revert "Remove duplicate definition"

This reverts commit 4fe0b0cba8eddc4992ae3adde759a6d57ac85cbf.

goggle committed:
Revert "Combine `Dense{T}` methods into one"

This reverts commit 9d76ccff56e22f82ca9808495a3857a8f30da807.

dkarrasch committed:
Merge branch 'master' into 27860-cholmod

goggle commented:
@dkarrasch You seem to have pushed something onto my branch. Was this an accident?
Also, is this ready to be merged?

dkarrasch committed:
fix typo in test

dkarrasch commented:
Oh yes, I resolved the merge conflict, with the intention to merge right afterwards... and forgot! 🤣 To make matters worse, I have introduced an issue by missing an `end` somewhere in the `cholmod` test files. 😇 I'm sorry, I didn't want to bother you with the "trivial" merge conflict. I quickly fixed it now, and will merge after tests pass.
